File: external/hypergraph-of-entity/src/main/java/armyant/hgoe/structures/ResultSet.java
package armyant.hgoe.structures;

import org.jetbrains.annotations.NotNull;

import java.util.*;

/**
 * Created by jldevezas on 2017-11-07.
 */
public class ResultSet implements Iterator<Result>, Iterable<Result> {
    private SortedSet<Result> results;
    private Map<String, Result> maxResultPerDocID;
    private Long numDocs;
    private Trace trace;
    private Iterator<Result> resultsIterator;

    public ResultSet() {
        this(new TreeSet<>((a, b) -> Double.compare(b.getScore(), a.getScore())), null, null);
    }

    public ResultSet(SortedSet<Result> results) {
        this(results, null, null);
    }

    public ResultSet(SortedSet<Result> results, Long numDocs, Trace trace) {
        this.results = results;
        this.maxResultPerDocID = new HashMap<>();
        addReplaceResults(results);
        this.numDocs = numDocs;
        this.trace = trace;
        this.resultsIterator = null;
    }

    public SortedSet<Result> getResults() {
        return results;
    }

    public void setResults(TreeSet<Result> results) {
        this.results = results;
    }

    public Long getNumDocs() {
        if (numDocs == null) {
            numDocs = (long) results.size();
            return numDocs;
        }
        return numDocs;
    }

    public void setNumDocs(Long numDocs) {
        this.numDocs = numDocs;
    }

    public Trace getTrace() {
        return trace;
    }

    public void setTrace(Trace trace) {
        this.trace = trace;
    }

    public void unsetTrace() {
        this.trace = null;
    }

    public void addResult(Result result) {
        results.add(result);
    }

    public void addReplaceResult(Result result) {
        maxResultPerDocID.put(result.getDocID(), result);
        results.add(result);

        Result maxResult = maxResultPerDocID.get(result.getDocID());
        if (maxResult != null && result.getScore() > maxResult.getScore()) {
            results.remove(maxResult);
        }
    }

    public void addResults(SortedSet<Result> results) {
        this.results.addAll(results);
    }

    public void addReplaceResults(SortedSet<Result> results) {
        for (Result result : results) {
            addReplaceResult(result);
        }
    }

    public boolean removeResult(Result result) {
        return results.remove(result);
    }

    public boolean removeResults(List<Result> results) {
        return this.results.removeAll(results);
    }

    @Override
    public boolean hasNext() {
        if (resultsIterator == null) {
            resultsIterator = results.iterator();
        }
        return resultsIterator.hasNext();
    }

    @Override
    public Result next() {
        return resultsIterator.next();
    }

    public static ResultSet empty() {
        return new ResultSet();
    }

    @NotNull
    @Override
    public Iterator<Result> iterator() {
        return results.iterator();
    }
}

File: external/cpp-impl/src/structures/result_set.cpp
//
// Created by jldevezas on 3/16/18.
//

#include <structures/result_set.h>

ResultSet::ResultSet(const ResultSet::results_t &results) : results(results), numDocs(nullptr) {}

ResultSet::ResultSet(const ResultSet::results_t &results, unsigned int numDocs) : results(results), numDocs(&numDocs) {}

ResultSet::iterator ResultSet::begin() {
    return results.begin();
}

ResultSet::iterator ResultSet::end() {
    return results.end();
}

ResultSet::const_iterator ResultSet::begin() const {
    return results.begin();
}

ResultSet::const_iterator ResultSet::end() const {
    return results.end();
}

ResultSet::const_iterator ResultSet::cbegin() const {
    return results.cbegin();
}

ResultSet::const_iterator ResultSet::cend() const {
    return results.cend();
}

const ResultSet::results_t &ResultSet::getResults() const {
    return results;
}

void ResultSet::setResults(const ResultSet::results_t &results) {
    ResultSet::results = results;
}

unsigned int ResultSet::getNumDocs() const {
    if (numDocs == nullptr) {
        numDocs = boost::make_shared<unsigned int>(results.size());
    }
    return *numDocs;
}

void ResultSet::setNumDocs(const unsigned int numDocs) {
    ResultSet::numDocs = boost::make_shared<unsigned int>(numDocs);
}

void ResultSet::addResult(Result result) {
    results.insert(result);
}

void ResultSet::addReplaceResult(Result result) {
    maxResultPerDocID[result.getDocID()] = result;
    results.insert(result);

    auto maxResult = maxResultPerDocID.find(result.getDocID());
    if (maxResult != maxResultPerDocID.end() && result.getScore() > (*maxResult)->getScore()) {
        results.erase(maxResult);
    }
}